### DISCUSSION TITLE
How to disconnect by bless

### DISCUSSION THREAD

**Discussion by trueroad** [kevincar/bless#97] [Q&A]
**Describe the solution you'd like**
I would like to disconnect by bless (i.e., by peripheral), not by central.

If I understand correctly,
`BlessServer.stop()` stops advertising but does not disconnect.
The destructor of `BlessServer` class also does not disconnect.
Therefore, unless disconnected by central, the connection keeps to remain even if the Python script that imports bless exits.

**Describe alternatives you've considered**
`BleakClient` class has `disconnect()` method.
But, `BlessServer` class does not have the `disconnect()` method.

**Additional context**
I use Windows 10.

**Comment by kevincar** [comment]
My understanding is that what you're describing is inherent to how Central and Peripheral devices are intended to function. Peripherals serve data to Central devices, while central devices connect to and read/write to peripheral data. Peripherals control advertising, while clients control the connection.

Because multiple clients could be connected at one time, it would be difficult for the peripheral to know whether it should disconnect all or just a specific one. Even if you got a handle on the device you wanted to disconnect, attempting to close the connection would depend on whether the device has no other app connections (See [here](https://learn.microsoft.com/en-us/uwp/api/windows.devices.bluetooth.bluetoothledevice.close?view=winrt-22621))

If you truly need to stop all connections, you could consider shutting down the bluetooth module.

You could also consider implementing a protocol, where the `BlessServer` notifies the `BleakClient` of some command that then tells the `BleakClient` to disconnect.

Let me know if I'm way off here or if there is more to this than I'm understanding.

**Comment by trueroad** [comment]
> Because multiple clients could be connected at one time, it would be difficult for the peripheral to know whether it should disconnect all or just a specific one. Even if you got a handle on the device you wanted to disconnect, attempting to close the connection would depend on whether the device has no other app connections (See [here](https://learn.microsoft.com/en-us/uwp/api/windows.devices.bluetooth.bluetoothledevice.close?view=winrt-22621))

In my usage, the script by python-bless is the only application connected to the central device.
Is there a way to enumerate the devices connected in `BlessServer` ?
Furthermore, is there a way to call the `BluetoothLEDevice.Close` method in `BlessServer` ?

> If you truly need to stop all connections, you could consider shutting down the bluetooth module.

Is there a way to implement shutting down and re-start the Bluetooth module in Bless?

> You could also consider implementing a protocol, where the BlessServer notifies the BleakClient of some command that then tells the BleakClient to disconnect.

I use WIDI Master which is BLE-MIDI adapter as the central device.
It uses the BLE-MIDI protocol that has no disconnect command.

**Comment by kevincar** [comment]
> Is there a way to enumerate the devices connected in BlessServer ?

At the moment, it cannot enumerate connected central devices

> is there a way to call the `BluetoothLEDevice.Close` method in `BlessServer`?

I'm assuming you're referring to [this method](https://learn.microsoft.com/en-us/uwp/api/windows.devices.bluetooth.bluetoothledevice.close?view=winrt-22621) on Microsoft's Bluetooth API? [A quick search on their sample code](https://github.com/microsoft/Windows-universal-samples/search?q=BluetoothLEDevice) shows that this method is typically used by client (central) devices, not peripheral devices. 

So currently there isn't a way to call this from `Bless` because, to my knowledge, there's no clear way to call `BluetoothLEDevice.close` as a peripheral. Though I'm hoping to dive into this more.

> Is there a way to implement shutting down and re-start the Bluetooth module in Bless?

Currently, no. Though I'm open to exploring this.

> It uses the BLE-MIDI protocol that has no disconnect command.

Is your script attempting to control the standard bluetooth specification of the BLE-MIDI or are you attempting to use the BLE-MIDI specification to connect to your WIDI Master? From the [BLE-MIDI Specification](https://www.midi.org/specifications-old/item/bluetooth-le-midi), it looks like the WIDI functions as a bluetooth peripheral device (aka server) which broadcasts its information.

If you're attempting to connect to the WIDI over python, you should be using [Bleak](https://github.com/hbldh/bleak) as you pointed out earlier.

Since this doesn't seem to be a direct issue with bless. I'm converting this to a discussion so we can continue hashing out the problems you're having. If there really is a lack on the Bless side or a potential feature implementation, I'll open this back up as an issue.

**Reply by trueroad** [reply]
WIDI Master has a mode that operates as a BLE-MIDI central.
It automatically connects when it finds a BLE-MIDI peripheral.

My BLE-MIDI peripheral script by Bless is here.
https://gist.github.com/trueroad/aa2507bf64c97f34aab324d8278b5686
When the script is started, WIDI Master automatically connects to the script.

However, WIDI Master remains connected even after the script is terminated.
And then I run the script again, WIDI Master does not re-connect to the script.

**Reply by kevincar** [reply]
Are there any negative consequences if the WIDI Master (BLE-MIDI central) remains connected to the peripheral emulator? If you start up another peripheral emulator, is the WIDI Master still able to find and auto-connect to that peripheral?

**Reply by trueroad** [reply]
WIDI Master (BLE-MIDI central) seems to use only one connection at a time.
That is, it cannot connect to other peripherals if it has the connection remained.

**Reply by kevincar** [reply]
What is the operating system you are using to run the `bless` peripheral? 

How are you able to confirm which peripheral the WIDI Master is connected to?

**Reply by trueroad** [reply]
I use Windows 10.

> How are you able to confirm which peripheral the WIDI Master is connected to?

When MIDI messages are sent from a MIDI keyboard to the WIDI Master by a MIDI cable, the BLE-connected peripheral gets the very same messages.
Unconnected peripherals do not get messages.

**Comment by pe224** [comment]
I'd like to +1 the request for enumerating connected central devices and selectively terminating the connections.
I think there are definitely use cases for killing the connection on the central side.

One example I'm struggling with is hardware testing. I'm using bless to mock a BLE central and I would like to simulate a connection drop.
But even shutting down the python process which started the bless server keeps the connection alive.

If you have some ideas where to look for this functionality (bleak package?), I could try. The tricky part will probably be the cross-platform aspect of it - I'm mainly on Windows and Linux without access to a Mac.

**Reply by pe224** [reply]
As a crude workaround, I now globally disable and re-enable the Bluetooth adapter, as described [here](https://stackoverflow.com/a/66418948/3764699).
This forcefully disconnects all peripherals - of course along with anything else which uses Bluetooth on this machine at that time.
A solution to terminate only selected connections would still be a lot cleaner.
